doom/config.org
#+title: Config
#+property: header-args :tangle config.el
#+auto_tangle: t
#+startup: overview

* Intro
Place your private configuration here! Remember, you do not need to run 'doom
sync' after modifying this file!


Some functionality uses this to identify you, e.g. GPG configuration, email
clients, file templates and snippets. It is optional.

Doom exposes five (optional) variables for controlling fonts in Doom:
- `doom-font' -- the primary font to use
- `doom-variable-pitch-font' -- a non-monospace font (where applicable)
- `doom-big-font' -- used for `doom-big-font-mode'; use this for
  presentations or streaming.
- `doom-unicode-font' -- for unicode glyphs
- `doom-serif-font' -- for the `fixed-pitch-serif' face

See 'C-h v doom-font' for documentation and more examples of what they
accept. For example:

setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
     doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))

If you or Emacs can't find your font, use 'M-x describe-font' to look them
up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
refresh your font settings. If Emacs still can't find your font, it likely
wasn't installed correctly. Font issues are rarely Doom issues!

There are two ways to load a theme. Both assume the theme is installed and
available. You can either set `doom-theme' or manually load a theme with the
`load-theme' function. This is the default:

This determines the style of line numbers in effect. If set to `nil', line
numbers are disabled. For relative line numbers, set this to `relative'.

If you use `org' and don't want your org files in the default location below,
change `org-directory'. It must be set before org loads!


Whenever you reconfigure a package, make sure to wrap your config in an
`after!' block, otherwise Doom's defaults may override your settings. E.g.

  (after! PACKAGE
    (setq x y))

The exceptions to this rule:

  - Setting file/directory variables (like `org-directory')
  - Setting variables which explicitly tell you to set them before their
    package is loaded (see 'C-h v VARIABLE' to look up their documentation).
  - Setting doom variables (which start with 'doom-' or '+').

Here are some additional functions/macros that will help you configure Doom.

- `load!' for loading external *.el files relative to this one
- `use-package!' for configuring packages
- `after!' for running code after a package has loaded
- `add-load-path!' for adding directories to the `load-path', relative to
  this file. Emacs searches the `load-path' when you load packages with
  `require' or `use-package'.
- `map!' for binding new keys

To get information about any of these functions/macros, move the cursor over
the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
This will open documentation for it, including demos of how they are used.
Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
etc).

You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
they are implemented.

* User
#+begin_src elisp
(setq user-full-name "Dean Gao"
      user-mail-address "gao.dean@hotmail.com")
#+end_src

* Startup benchmark
#+begin_src elisp
;; (require 'benchmark-init)
;; (add-hook 'doom-first-input-hook #'benchmark-init/deactivate)
#+end_src

* System
#+begin_src elisp
(setenv "PATH" (concat (getenv "PATH") ":/usr/local/texlive/2023/bin/universal-darwin"))
(setq exec-path (append exec-path '("/usr/local/texlive/2023/bin/universal-darwin")))
#+end_src

* Appearance
** Theme
#+begin_src elisp
(setq doom-theme 'doom-moonlight)
;; (setq doom-theme 'gruber-darker)

;; (custom-set-faces!
;;   `(default :background "#282c34"))
#+end_src

** No Titlebar
#+begin_src elisp
(add-to-list 'default-frame-alist '(undecorated . t))
#+end_src
** Transparent background
This is to make it blurry and transparent
#+begin_src elisp
;; (set-face-background 'default "mac:windowBackgroundColor")
;; (set-face-stipple 'default "alpha:80%")
#+end_src
** Scrolling
#+begin_src elisp
(pixel-scroll-precision-mode)
#+end_src
* Fonts
Settings related to fonts within Doom Emacs:
+ 'doom-font' -- standard monospace font that is used for most things in Emacs.
+ 'doom-variable-pitch-font' -- variable font which is useful in some Emacs plugins.
+ 'doom-big-font' -- used in doom-big-font-mode; useful for presentations.
+ 'font-lock-comment-face' -- for comments.
+ 'font-lock-keyword-face' -- for keywords with special significance like 'setq' in elisp.

#+begin_src elisp
(setq doom-font (font-spec :family "Input" :size 20)
      doom-variable-pitch-font (font-spec :family "Merriweather" :size 20)
      doom-big-font (font-spec :family "Input" :size 26))
(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))
(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))

#+end_src

* Org Mode
** org-auto-tangle
=org-auto-tangle= allows you to add the option =#+auto_tangle: t= in your Org file so that it automatically tangles when you save the document.  I have made adding this to your file even easier by creating a function 'dt/insert-auto-tangle-tag' and setting it to a keybinding 'SPC i a'.

#+begin_src elisp
(use-package! org-auto-tangle
  :defer t
  :hook (org-mode . org-auto-tangle-mode)
  :config
  (setq org-auto-tangle-default nil))

(defun dg/insert-auto-tangle-tag ()
  "Insert auto-tangle tag in a literate config."
  (interactive)
  (evil-org-open-below 1)
  (insert "#+auto_tangle: t ")
  (evil-force-normal-state))

(map! :leader
      :desc "Insert auto_tangle tag" "i a" #'dg/insert-auto-tangle-tag)
#+end_src

** org-autolist
Makes list keybindings function more like usual word processors

#+begin_src elisp
(add-hook 'org-mode-hook 'org-autolist-mode)
#+end_src
** org-appear
#+begin_src elisp
(add-hook 'org-mode-hook 'org-appear-mode)
(setq org-appear-autolinks t
      org-appear-autosubmarkers t
      org-appear-autoentities t
      org-appear-autokeywords t
      org-appear-inside-latex t
      org-appear-delay 0.1)
#+end_src

** org-fragtog
Latex preview
#+begin_src elisp
(setq org-startup-with-latex-preview t)
(setq org-latex-create-formula-image-program 'dvisvgm)
(after! org
  (plist-put org-format-latex-options :scale 2.6))
#+end_src

#+RESULTS:
| :foreground | default | :background | default | :scale | 2.6 | :html-foreground | Black | :html-background | Transparent | :html-scale | 1.0 | :matchers | (begin $1 $ $$ \( \[) |

*** Background colour
So the background color isn't darker: https://stackoverflow.com/questions/17239273/org-mode-buffer-latex-syntax-highlighting
#+begin_src elisp
(after! org
  (setq org-highlight-latex-and-related '(latex script entities)))
#+end_src
*** Tables
#+begin_src elisp
(after! org-fragtog (org-fragtog-table-mode))
#+end_src

** ob-mermaid
#+begin_src elisp
(setq ob-mermaid-cli-path "/Users/deangao/.local/share/npm/bin/mmdc")
(org-babel-do-load-languages
 'org-babel-load-languages
 '((mermaid . t)
   (latex . t)
   (emacs-lisp . t)))
(setq org-babel-default-header-args:mermaid
      '((:results . "file")
        (:exports . "results")
        (:puppeteer-config-file . "/Users/deangao/.config/puppeteer/config.json")))
#+end_src
** Variables
I'm using FiraCode for the superstar header bullet because the bullet isn't as big in FiraCode and looks better

#+begin_src elisp
(after! org
  (setq org-directory "~/des/"
        org-ellipsis " ▼ "
        org-pretty-entities t
        org-superstar-item-bullet-alist '((?- . ?➤) (?+ . ?✦)) ; changes +/- symbols in item lists
        org-log-done 'time
        org-hide-emphasis-markers t
        org-table-convert-region-max-lines 20000)
(display-line-numbers-mode nil))

(custom-set-faces!
  `(org-superstar-header-bullet :font "FiraCode NF" :height 1.1 :weight light)
  `(org-ellipsis :height 0.8 :underline nil :foreground "#484848"))
#+end_src

** Header size
#+begin_src elisp
(defun dg/set-org-header-size ()
  (interactive)
  (with-eval-after-load 'org-faces (dolist
      (face
       '((org-level-1 1.4 "#51afef" ultra-bold)
         (org-level-2 1.3 "#c678dd" extra-bold)
         (org-level-3 1.2 "#98be65" bold)
         (org-level-4 1.1 "#da8548" semi-bold)
         (org-level-5 1.0 "#5699af" normal)
         (org-level-6 1.0 "#a9a1e1" normal)
         (org-level-7 1.0 "#46d9ff" normal)
         (org-level-8 1.0 "#ff6c6b" normal)))
    (set-face-attribute (nth 0 face) nil :font "Source Sans Pro" :weight (nth 3 face) :height (nth 1 face) :foreground (nth 2 face))))
  (with-eval-after-load 'org-tables (set-face-attribute 'org-table nil :font doom-font :weight 'normal :height 1.0 :foreground "#bfafdf")))

(dg/set-org-header-size)
#+end_src
** Export
#+begin_src elisp
(after! org
  (setq org-latex-pdf-process (list "latexmk -f -pdfxe -interaction=nonstopmode -output-directory=%o %f")
        org-latex-default-packages-alist
        '(("AUTO" "inputenc" nil
          ("pdflatex"))
         ("T1" "fontenc" nil
          ("pdflatex"))
         ("" "graphicx" t)
         ("" "longtable" t)
         ("" "wrapfig" nil)
         ("" "rotating" nil)
         ("normalem" "ulem" t)

emacs/history
;; -*- mode: emacs-lisp; coding: utf-8-unix -*-
;; Minibuffer history file, automatically generated by ‘savehist’.

(setq savehist-minibuffer-history-variables '(minibuffer-history extended-command-history file-name-history evil-ex-history))
(setq minibuffer-history '("straight-check-for-modifications" "~/.config/emacs/config.org" "yes-or-no-p"))
(setq extended-command-history '("recentf"))
(setq file-name-history '("~/.config/emacs/config.org"))

emacs/config.el
(setq straight-check-for-modifications (list 'check-on-save 'find-when-checking))
(setq straight-repository-branch "develop")

(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 6))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

;; (straight-use-package 'use-package) ;; emacs 29 built-in use-package
(require 'use-package)
(setq straight-use-package-by-default t)

(setq straight-host-usernames
      '((github . "gaoDean")
        (gitlab . "gaoDean")))

(straight-use-package 'benchmark-init)
(require 'benchmark-init)
(add-hook 'after-init-hook 'benchmark-init/deactivate)

(use-package general)
(general-create-definer leader-def
                        ;; :prefix my-leader
                        :prefix "SPC")

(use-package evil
  :init
  (setq evil-want-integration t
        evil-want-keybinding nil
        evil-want-minibuffer t
        evil-undo-system 'undo-fu)
  :config
  (evil-mode 1))

(use-package evil-collection
  :after evil
  :config
  (evil-collection-init))

(use-package evil-surround
  :config
  (global-evil-surround-mode 1))

(use-package evil-org
  :hook org-mode)

(use-package evil-numbers
  :config
  (leader-def :keymaps '(normal visual) "a +" 'evil-numbers/inc-at-pt)
  (leader-def :keymaps '(normal visual) "a -" 'evil-numbers/dec-at-pt))

(use-package evil-commentary
  :config
  (evil-commentary-mode))

(setq user-full-name "Dean Gao"
      user-mail-address "gao.dean@hotmail.com")

(add-to-list 'load-path "~/.config/emacs/plugins/")

(use-package nano-theme
  :straight (nano-theme :type git :host github :repo "rougier/nano-theme")
  :custom-face
  (default ((t (:family "Input Mono" :height 240))))
  (italic ((t (:family "Input Mono" :height 240 :slant italic))))
  (variable-pitch ((t (:family "Lato" :height 240))))
  :config
  (nano-dark))

(use-package nano-splash
  :custom
  (nano-splash-duration 20)
  :straight (nano-splash :type git :host github :repo "gaoDean/nano-splash")
  :config (nano-splash))

(with-eval-after-load 'nano-splash
  (require 'nano-session))

(defun mode-line-render (left right)
  (let* ((available-width (- (window-width) (length left) )))
    (format (format "%%s %%%ds" available-width) left right)))
(setq-default header-line-format
              '((:eval
                 (mode-line-render
                  (format-mode-line (list
                                     (propertize "☰" 'face `(:inherit mode-line-buffer-id)
                                                 'help-echo "Mode(s) menu"
                                                 'mouse-face 'mode-line-highlight
                                                 'local-map   mode-line-major-mode-keymap)
                                     " %b "
                                     (if (and buffer-file-name (buffer-modified-p))
                                         (propertize "[M]" 'face `(:inherit nano-faded)))))
                  (format-mode-line (propertize "%4l:%2c  " 'face `(:inherit nano-faded)))))))
(setq-default mode-line-format nil)
(set-face-attribute 'header-line nil
                    :underline nil
                    :background nil)

(set-face-background 'header-line nil)

(use-package mixed-pitch
  :hook text-mode)

(use-package all-the-icons)

(use-package writeroom-mode
  :custom
  (writeroom-fullscreen-effect 'maximized)
  (writeroom-header-line t)
  :general
  (leader-def :keymaps 'normal "t z" 'writeroom-mode))

(setq-default fill-column 81 ;; Thou shalt not cross 80 columns in thy file
                sentence-end-double-space nil           ; Use a single space after dots
                bidi-paragraph-direction 'left-to-right ; Faster
                truncate-string-ellipsis "…")           ; Nicer ellipsis

  (setq x-underline-at-descent-line nil
        x-use-underline-position-properties t
        underline-minimum-offset 10)

  (add-hook 'text-mode-hook 'visual-line-mode)
  (add-hook 'prog-mode-hook 'visual-line-mode)

(setq-default indent-tabs-mode nil        ; Stop using tabs to indent
              tab-always-indent 'complete ; Indent first then try completions
              tab-width 4)                ; Smaller width for tab characters

;; Let Emacs guess Python indent silently
(setq python-indent-guess-indent-offset t
      python-indent-guess-indent-offset-verbose nil)

(use-package helpful
  :general
  (leader-def 'normal
    "h F" 'describe-face
    "h p" 'describe-package
    "h f" 'helpful-callable
    "h b" 'describe-bindings
    "h v" 'helpful-variable
    "h k" 'helpful-key
    "h x" 'helpful-command
    "h ." 'helpful-at-point))

(use-package which-key
  :init
  (setq which-key-show-early-on-C-h t)
  (setq which-key-idle-delay 3)
  :hook emacs-startup
  :config
  (which-key-setup-side-window-right))

(setq ido-ignore-buffers '("^ " "\*"))

(use-package rg
  :commands (rg)
  :config
  (rg-enable-menu))

;; (use-package flycheck
;;   :hook (after-init . global-flycheck-mode))

(use-package hl-todo
  :hook emacs-startup)

(setq backup-directory-alist `(("." . ,"~/.cache/emacs/backups")))

(setq create-lockfiles nil)

(use-package undo-fu)
(use-package undo-fu-session
  :custom
  (undo-fu-session-directory "~/.cache/emacs/undo-fu-session")
  :hook (emacs-startup . undo-fu-session-global-mode))

(defun my/set-face (face inherit)
  (face-spec-reset-face face)
  (set-face-attribute face nil :inherit inherit))

(use-package whitespace-cleanup-mode
  :hook (prog-mode . whitespace-cleanup-mode))

(use-package whitespace
  :straight nil
  :init
  (setq whitespace-style '(face empty tabs lines-tail trailing))
  (setq whitespace-line-column 81) ;; Thou shalt not cross 80 columns in thy file
  :general
  (leader-def :keymaps 'normal "t w" 'whitespace-mode)
  :config
  (my/set-face 'whitespace-line 'nano-critical)
  (my/set-face 'whitespace-trailing 'nano-critical-i)
  (my/set-face 'whitespace-tab 'nano-critical-i))

(use-package fcitx
  :hook (emacs-startup . fcitx-aggressive-setup))

(use-package org :straight (:type built-in))

(use-package org-modern
  :init
  (setq
   org-catch-invisible-edits 'show-and-error
   org-insert-heading-respect-content t
   org-hide-emphasis-markers t
   org-modern-label-border 0.3
   org-modern-hide-stars " "
   org-image-actual-width '(400)
   line-spacing 0.1
   org-pretty-entities t
   org-ellipsis "…")
  :hook org-mode
  :config
  (with-eval-after-load 'org-faces
    (set-face-attribute 'org-level-1 nil :font "Source Sans Pro" :weight 'bold :height 1.4)
    (set-face-attribute 'org-level-2 nil :font "Source Sans Pro" :weight 'bold :height 1.3)
    (set-face-attribute 'org-level-3 nil :font "Source Sans Pro" :weight 'bold :height 1.2)
    (set-face-attribute 'org-level-4 nil :font "Source Sans Pro" :weight 'bold :height 1.1)
    (set-face-attribute 'org-level-5 nil :font "Source Sans Pro" :weight 'bold :height 1.0)
    (set-face-attribute 'org-level-6 nil :font "Source Sans Pro" :weight 'bold :height 1.0)
    (set-face-attribute 'org-level-7 nil :font "Source Sans Pro" :weight 'bold :height 1.0)
    (set-face-attribute 'org-modern-symbol nil :font "FiraCode NF" :height 1.1)))

(use-package org-autolist :hook org-mode)

(use-package org-fragtog
  :init
  (setq org-startup-with-latex-preview nil
        org-latex-create-formula-image-program 'dvisvgm
        org-highlight-latex-and-related '(latex script entities)
        org-latex-preview-ltxpng-directory "~/.cache/emacs/ltxpng/")
  :config
  (plist-put org-format-latex-options :scale 2.6)
  :hook org-mode)

(use-package org-appear
  :hook org-mode
  :custom
  (org-appear-autoentities t)
  (org-appear-submarkers t)
  (org-appear-autolinks t))

(use-package org-auto-tangle :hook org-mode)

(with-eval-after-load 'org
  (require 'ox-latex)

  (setq org-latex-pdf-process (list "latexmk -f -pdfxe -interaction=nonstopmode -output-directory=%o %f")
        org-latex-default-packages-alist
        '(("AUTO" "inputenc" nil
           ("pdflatex"))
          ("T1" "fontenc" nil
           ("pdflatex"))
          ("" "graphicx" t)
          ("" "longtable" t)
          ("" "wrapfig" nil)
          ("" "rotating" nil)
          ("normalem" "ulem" t)
          ("" "amsmath" t)
          ("" "amssymb" t)
          ("" "capt-of" nil)
          ("" "hyperref" t)))

(add-to-list 'org-latex-classes
             '("orgox"
               "
                \\documentclass[hidelinks]{article}
                [DEFAULT-PACKAGES]
                [PACKAGES]
                [EXTRA]"
               ("\\section{%s}" . "\\section*{%s}")
               ("\\subsection{%s}" . "\\subsection*{%s}")
               ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
               ("\\paragraph{%s}" . "\\paragraph*{%s}")
               ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))))

(use-package citeproc :if (eq major-mode 'org-mode))

(use-package htmlize)

(setq org-html-postamble-format '(("en" "
<p class=\"author\">Author: %a</p>
<p class=\"updated\">Updated: %C</p>
<p class=\"creator\">%c</p>
")))

(setq org-publish-timestamp-directory "~/.cache/emacs/org-timestamps/")  
(setq org-publish-project-alist
      '(("org-notes"
         :base-directory "~/org/"
         :publishing-directory "~/org/pub/"
         :base-extension "org"
         :publishing-function org-html-publish-to-html
         :exclude "pub"
         :recursive t
         :html-extension "html"
         :auto-preamble t
         :html-postamble t
         :section-numbers t
         :with-toc t
         :html-head "<link rel=\"stylesheet\"
                                   href=\"/web/main.css\"
                                   type=\"text/css\"/>")
        ("org-static"
         :base-directory "~/org/"
         :base-extension "css\\|js\\|png\\|jpg\\|gif\\|pdf\\|mp3\\|ogg\\|swf"
         :publishing-directory "~/org/pub/"
         :exclude "pub"
         :recursive t
         :publishing-function org-publish-attachment
         )
        ("org" :components ("org-notes" "org-static"))))

(use-package org-imgtog
  :custom
  (org-imgtog-preview-delay 0.2)
  :straight (org-imgtog :type git :host github :repo "gaoDean/org-imgtog" :local-repo "~/repos/rea/org-imgtog")
  :hook org-mode)

(setq url-cache-directory "~/.cache/emacs/url")
(setq org-display-remote-inline-images 'cache)
;; (setq org-display-remote-inline-images 'download)

(use-package org-remoteimg
  :straight (org-remoteimg :type git :host github :repo "gaoDean/org-remoteimg" :local-repo "~/repos/rea/org-remoteimg"))

(use-package avy
    :custom
    (avy-keys '(?i ?s ?r ?t ?g ?p ?n ?e ?a ?o))
    :general
    (:keymaps 'normal ";" 'avy-goto-char-2)
    (:keymaps '(insert visual) "C-;" 'avy-goto-char-2)
)

(use-package yaml-mode
  :mode ("\\.ya?ml$'" . yaml-mode))

(use-package lua-mode
  :init
  (setq lua-indent-level 4)
  :mode ("\\.lua$" . lua-mode))

(use-package magit
  :general
  (leader-def 'normal "g" 'magit))

(use-package projectile
    :config
    (projectile-mode)
    :general
    (leader-def 'normal
        "p p" 'projectile-switch-project
        "p g" 'projectile-ripgrep
        "SPC" 'projectile-find-file))

(use-package vertico
  :straight (:files (:defaults "extensions/*"))
  :init
  (setq vertico-resize nil        ; How to resize the Vertico minibuffer window.
        vertico-count 8           ; Maximal number of candidates to show.
        vertico-count-format nil)

  (setq vertico-grid-separator
        #("  |  " 2 3 (display (space :width (1))
                               face (:background "#ECEFF1")))

        vertico-group-format
        (concat #(" " 0 1 (face vertico-group-title))
                #(" " 0 1 (face vertico-group-separator))
                #(" %s " 0 4 (face vertico-group-title))
                #(" " 0 1 (face vertico-group-separator
                                display (space :align-to (- right (-1 . right-margin) (- +1)))))))
  (vertico-mode)

  :config
  (set-face-attribute 'vertico-group-separator nil
                      :strike-through t)
  (set-face-attribute 'vertico-current nil
                      :inherit '(nano-strong nano-subtle))
  (set-face-attribute 'completions-first-difference nil
                      :inherit '(nano-default))
  :general
  (:keymaps 'vertico-map
            "?" 'minibuffer-completion-help
            "C-j" 'vertico-next
            "C-k" 'vertico-previous
            "M-TAB" 'minibuffer-complete))

(use-package savehist
  :config
  (savehist-mode))

(use-package marginalia
  :init
  (setq-default marginalia--ellipsis "…"    ; Nicer ellipsis
                marginalia-align 'right     ; right alignment
                marginalia-align-offset -1) ; one space on the right
  :config
  (marginalia-mode))

(use-package orderless
  :init
  (setq completion-styles '(substring orderless basic)
        orderless-component-separator 'orderless-escapable-split-on-space
        read-file-name-completion-ignore-case t
        read-buffer-completion-ignore-case t
        completion-ignore-case t))

(use-package tempel
  :config
  ;; Setup completion at point
  (defun tempel-setup-capf ()
    (setq-local completion-at-point-functions
                (cons #'tempel-expand
                      completion-at-point-functions)))

  (add-hook 'prog-mode-hook 'tempel-setup-capf)
  (add-hook 'text-mode-hook 'tempel-setup-capf))

(use-package tempel-collection :after tempel)

(use-package corfu
  :straight (:files (:defaults "extensions/*"))
  :hook (emacs-startup . global-corfu-mode)
  :init
  (setq tab-always-indent 'complete)
  :config
  (defun corfu-enable-always-in-minibuffer ()
    (unless (bound-and-true-p vertico--input))
      (setq-local corfu-auto nil) 
      (corfu-mode 1))
  (add-hook 'minibuffer-setup-hook #'corfu-enable-always-in-minibuffer 1)
  (keymap-set corfu-map "M-q" #'corfu-quick-complete)
  (keymap-set corfu-map "C-q" #'corfu-quick-insert)
  :custom
  (corfu-cycle t)           ;; Enable cycling for `corfu-next/previous'
  (corfu-preselect-first t) ;; Always preselect the prompt
  (corfu-echo-delay '(1.0 0.5))
  :general
  (:keymaps 'corfu-map
            "TAB" 'corfu-next
            "S-TAB" 'corfu-previous))

(use-package cape
  :config

  (defun cape-setup-capf ()
    (add-to-list 'completion-at-point-functions #'cape-dabbrev)
    ;; (add-to-list 'completion-at-point-functions #'cape-file)
    (add-to-list 'completion-at-point-functions #'cape-elisp-block)
    ;; (add-to-list 'completion-at-point-functions #'cape-history)
    ;; (add-to-list 'completion-at-point-functions #'cape-keyword)
    ;; (add-to-list 'completion-at-point-functions #'cape-tex)
    ;; (add-to-list 'completion-at-point-functions #'cape-sgml)
    ;; (add-to-list 'completion-at-point-functions #'cape-rfc1345)
    ;; (add-to-list 'completion-at-point-functions #'cape-abbrev)
    ;; (add-to-list 'completion-at-point-functions #'cape-dict)
    ;; (add-to-list 'completion-at-point-functions #'cape-symbol)
    ;; (add-to-list 'completion-at-point-functions #'cape-line)
    )

  (add-hook 'prog-mode-hook 'cape-setup-capf)
  (add-hook 'text-mode-hook 'cape-setup-capf)

  )

(use-package shrink-path)
(use-package eshell-vterm
  :config
  (defalias 'eshell/v 'eshell-exec-visual))
(use-package eshell-up)

(add-hook 'eshell-mode-hook (lambda () (setenv "TERM" "xterm-256color")))

(defun fzf (strings)
  (completing-read "Filter: " strings))

(defun eshell/get-brew-formulae ()
  (split-string (shell-command-to-string "brew formulae") "\n" t))

(defun eshell/bi ()
                    (let ((selected (fzf (eshell/get-brew-formulae))))
                      (insert (concat "brew install " selected))
                      (eshell-command-result (concat "brew info " selected))))

(setq my/eshell-alises '(
                         ("f"  . "find-file")
                         ("l"  . "ls -lh $*")
                         ("la" . "ls -alh $*")
                         ("gs" . "magit-status $*")
                         ("g"  . "magit $*")
                         ("d"  . "dirvish $*")
                         ))



(add-hook 'eshell-mode-hook (lambda ()
                              (dolist (pair my/eshell-alises)
                                (eshell/alias (car pair) (cdr pair)))))

(setq eshell-prompt-regexp "^.* λ "
      eshell-prompt-function #'+eshell/prompt)

(defun +eshell/prompt ()
  (let ((base/dir (shrink-path-prompt default-directory)))
        (concat (propertize (car base/dir)
                            'face 'font-lock-comment-face)
                (propertize (cdr base/dir)
                            'face 'font-lock-constant-face)
                (propertize (+eshell--current-git-branch)
                            'face 'font-lock-function-name-face)
                (propertize " λ" 'face 'eshell-prompt-face)
                ;; needed for the input text to not have prompt face
                (propertize " " 'face 'default))))

;; for completeness sake
(defun +eshell--current-git-branch ()
    (let ((branch (car (cl-loop for match in (split-string (shell-command-to-string "git branch") "\n")
                             when (string-match "^\*" match)
                             collect match))))
      (if (not (eq branch nil))
          (concat " [" (substring branch 2) "]")
        "")))

(defun my/dired-up-directory-in-buffer ()
  (interactive)
  (find-alternate-file ".."))

(defun my/kill-all-dired-buffers-and-quit ()
"Kill all Dired buffers and quit the current Dired buffer."
(interactive)
(quit-window)
(mapc (lambda (buffer)
        (when (eq 'dired-mode (buffer-local-value 'major-mode buffer))
          (kill-buffer buffer)))
      (buffer-list)))


(use-package dired-open :after dirvish)

(use-package dirvish
  :hook (emacs-startup . dirvish-override-dired-mode)
  :straight (dirvish :type git :host github :repo "isamert/dirvish")
  :custom
  (dirvish-quick-access-entries ; It's a custom option, `setq' won't work
   '(("h" "~/"                          "Home")
     ("d" "~/Downloads/"                "Downloads")
     ("v" "~/vau/"                      "vau")
     ("r" "~/repos/"                    "repos")
     ("t" "~/.Trash"                    "Trash")))
  :init
  ;; (dirvish-peek-mode) ; Preview files in minibuffer
  ;; (dirvish-side-follow-mode) ; similar to `treemacs-follow-mode'
  (put 'dired-find-alternate-file 'disabled nil)
  (setq dirvish-mode-line-format '(:left (sort symlink) :right (omit yank index))
        dirvish-attributes '(all-the-icons file-time file-size collapse subtree-state vc-state git-msg)
        dired-recursive-copies 'always
        dired-recursive-deletes 'always
        delete-by-moving-to-trash t
        insert-directory-program "gls"
        dired-kill-when-opening-new-dired-buffer t
        dired-use-ls-dired t
        dired-listing-switches
        "-l --almost-all --human-readable --group-directories-first --no-group")
  :config
  (evil-define-key 'normal dired-mode-map
    (kbd "% l") 'dired-downcase
    (kbd "% m") 'dired-mark-files-regexp
    (kbd "% u") 'dired-upcase
    (kbd "* %") 'dired-mark-files-regexp
    (kbd "* .") 'dired-mark-extension
    (kbd "* /") 'dired-mark-directories
    (kbd "+") 'dired-create-directory
    (kbd "-") 'dirvish-narrow
    (kbd "<tab>") 'dirvish-toggle-subtree
    (kbd "M") 'dirvish-mark-menu
    (kbd "S") 'dirvish-symlink
    (kbd "a") 'dirvish-quick-access
    (kbd "c") 'dirvish-chxxx-menu
    (kbd "d") 'dired-do-delete
    (kbd "x") 'dired-do-delete
    (kbd "f") 'dirvish-file-info-menu
    (kbd "h") 'dired-up-directory
    (kbd "l") 'dired-find-alternate-file
    (kbd "o") 'dired-open-file
    (kbd "q") 'my/kill-all-dired-buffers-and-quit
    (kbd "m") 'dired-mark
    (kbd "p") 'dirvish-yank
    (kbd "r") 'dired-do-rename
    (kbd "t") 'find-file
    (kbd "u") 'dired-unmark
    (kbd "v") 'dirvish-move
    (kbd "y") 'dirvish-yank-menu
    (kbd "z") 'dired-do-compress))

(setq delete-by-moving-to-trash t
      trash-directory "~/.Trash")

(set-register ?c (cons 'file "~/.config/emacs/config.org"))
(set-register ?i (cons 'file "~/org/index.org"))

(defun my/reload-init-file ()
  (interactive)
  (load-file user-init-file))

(defun my/view-with-quicklook ()
  (interactive)
  (let* ((current-file (file-name-nondirectory (buffer-file-name)))
         (file-extensions '(".pdf" ".html"))
         (found-file nil)
         (file-extension ""))
    (while (and file-extensions (not found-file))
      (setq file-extension (car file-extensions))
      (let ((file-path (concat (file-name-directory (buffer-file-name))
                               (file-name-sans-extension current-file)
                               file-extension)))
        (when (file-exists-p file-path)
          (setq found-file file-path)))
      (setq file-extensions (cdr file-extensions)))
    (if found-file
        (async-start-process "quicklook" "qlmanage" nil "-p " found-file)
      (message "Exported file not found"))))

(defun my/git-push-org ()
  (interactive)
  (org-publish "org")
  (start-process "git-push-org" nil "gaa" "org")
  (message "done"))

(unbind-key "s-p") ;; ns-print-buffer

(leader-def :keymaps 'normal
  "b" '(:ignore t :wk "buffers")
  "b b" 'ido-switch-buffer
  "b B" 'ibuffer
  "b K" 'ido-kill-buffer
  "b k" 'kill-this-buffer
  "b n" 'bs-cycle-next
  "b p" 'bs-cycle-previous

  "w" '(:ignore t :wk "windows")
  "w w" 'evil-window-next
  "w k" 'evil-window-up
  "w j" 'evil-window-down
  "w h" 'evil-window-left
  "w l" 'evil-window-right
  "w p" 'evil-window-mru
  "w c" 'evil-window-delete
  "w v" 'evil-window-vsplit
  "w >" 'evil-window-increase-width
  "w <" 'evil-window-decrease-width
  "w +" 'evil-window-increase-height
  "w -" 'evil-window-decrease-height

  "o" '(:ignore t :wk "open")
  "o e" 'eshell
  "o v" 'vterm

  "t" '(:ignore t :wk "toggle")

  "a" '(:ignore t :wk "actions")
  "a e" 'org-export-dispatch
  "a p" 'org-publish
  "a o" 'my/git-push-org

  "f" '(:ignore t :wk "files")
  "f r" 'recentf
  "f v" 'my/view-with-quicklook

  "e" '(:ignore t :wk "emacs")
  "e r" 'my/reload-init-file
  "e m" 'toggle-frame-maximized

  "e e" '(:ignore t :wk "eval")
  "e e r" 'eval-region
  "e e b" 'eval-buffer
  "e e f" 'eval-defun

  "e b" '(:ignore t :wk "benchmark")
  "e b t" 'benchmark-init/show-durations-tabulated
  "e b r" 'benchmark-init/show-durations-tree
  "e b i" 'emacs-init-time

  "c" '(:ignore t :wk "code")
  "c w" 'fill-paragraph

  "."   'find-file
  "q" 'save-buffers-kill-terminal
  "s" 'scratch-buffer
  "r" 'jump-to-register)

(general-define-key :keymaps 'insert "<backtab>" 'evil-shift-left-line)

(general-define-key :keymaps 'insert "DEL" 'backward-delete-char-untabify)

(general-define-key :states '(normal visual) :keymaps 'override
                    "g j" 'evil-next-visual-line
                    "g k" 'evil-previous-visual-line
                    "C-u" 'evil-scroll-up)

(general-define-key "M-v" 'evil-paste-after)

(general-define-key (kbd "C-x C-m") 'execute-extended-command)
